go/vt/tabletserver/planbuilder/dml.go

```diff
@@ -323,12 +328,40 @@ func analyzeInsert(ins *sqlparser.Insert, getTable TableGetter) (plan *ExecPlan,
 	upd := &sqlparser.Update{
 		Comments: ins.Comments,
 		Table:    ins.Table,
-		Exprs:    sqlparser.UpdateExprs(ins.OnDup),
+		Exprs:    updateExprs[0],
 	}
 	plan.UpsertQuery = GenerateUpdateOuterQuery(upd)
 	return plan, nil
 }
 
+// generateUpsertUpdateExprs creates new UpdateExprs with any VALUES()
+// references replaced with the value they were meant to lookup
+// i.e. VALUES(foo) lookups up the insert value for column foo from the original
+// insert rows.
+// Supports multi-row inserts, though this is disallowed in analyzeInsert above
+func generateUpsertUpdateExprs(tableInfo *schema.Table, rowList sqlparser.Values, dup sqlparser.OnDup) ([]sqlparser.UpdateExprs, error) {
+	result := make([]sqlparser.UpdateExprs, len(rowList))
+	for i, rowTuple := range rowList {
+		rowResults := make([]*sqlparser.UpdateExpr, len(dup))
+		for j, update := range dup {
+			colResult := &sqlparser.UpdateExpr{Name: update.Name}
+			if update.UseLookup {
+				colID := tableInfo.FindColumn(update.LookupName)
```

Comment:
In response to your search question, linear searches should actually be more efficient than a map for a handful of columns, because we avoid the cost of computing the hash.